crates/tokio-util/src/event_listeners.rs

```diff
@@ -1,46 +1,59 @@
-use tokio::sync::mpsc;
-use tokio_stream::wrappers::UnboundedReceiverStream;
+use tokio::sync::broadcast::{self, Sender};
+use tokio_stream::wrappers::BroadcastStream;
+use tracing::{error, warn};
 
-/// A collection of event listeners for a task.
-#[derive(Clone, Debug)]
+const DEFAULT_BROADCAST_CHANNEL_SIZE: usize = 1000;
+
+/// A bounded broadcast channel for a task.
+#[derive(Debug)]
 pub struct EventListeners<T> {
-    /// All listeners for events
-    listeners: Vec<mpsc::UnboundedSender<T>>,
+    /// The sender part of the broadcast channel
+    sender: Sender<T>,
+}
+
+impl<T: Clone> Clone for EventListeners<T> {
+    fn clone(&self) -> Self {
+        EventListeners { sender: self.sender.clone() }
+    }
 }
 
-impl<T> Default for EventListeners<T> {
+impl<T: Clone + Send + Sync + 'static> Default for EventListeners<T> {
     fn default() -> Self {
-        Self { listeners: Vec::new() }
+        Self::new(DEFAULT_BROADCAST_CHANNEL_SIZE)
     }
 }
 
-impl<T: Clone> EventListeners<T> {
-    /// Send an event to all listeners.
-    ///
-    /// Channels that were closed are removed.
-    pub fn notify(&mut self, event: T) {
-        self.listeners.retain(|listener| listener.send(event.clone()).is_ok())
+impl<T: Clone + Send + Sync + 'static> EventListeners<T> {
+    /// Creates a new `EventListeners`.
+    pub fn new(broadcast_channel_size: usize) -> Self {
+        let (sender, _) = broadcast::channel(broadcast_channel_size);
+        Self { sender }
     }
 
-    /// Add a new event listener.
-    pub fn new_listener(&mut self) -> UnboundedReceiverStream<T> {
-        let (sender, receiver) = mpsc::unbounded_channel();
-        self.listeners.push(sender);
-        UnboundedReceiverStream::new(receiver)
+    /// Broadcast sender setter.
+    pub fn set_sender(&mut self, sender: Sender<T>) {
+        self.sender = sender;
     }
 
-    /// Push new event listener.
-    pub fn push_listener(&mut self, listener: mpsc::UnboundedSender<T>) {
-        self.listeners.push(listener);
+    /// Broadcasts an event to all listeners.
+    pub fn notify(&self, event: T) {
+        match self.sender.send(event) {
+            Ok(listener_count) => {
+                if listener_count == 0 {
+                    warn!("notification of network event with 0 listeners");
+                }
```

Comment:
this wouldn't be zero I think, then an error would be thrown as it would mean all receivers are dropped https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Sender.html#method.send

Reply:
good catch! fixed